flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/SpillingThread.java

```diff
@@ -0,0 +1,601 @@
+			IOManager ioManager,
+			TypeSerializer<E> serializer,
+			TypeComparator<E> comparator,
+			List<MemorySegment> sortReadMemory,
+			List<MemorySegment> writeMemory,
+			int maxNumFileHandles,
+			SpillChannelManager spillingChannelManager,
+			LargeRecordHandler<E> largeRecordHandler,
+			SpillingBehaviour<E> spillingBehaviour,
+			int minNumWriteBuffers,
+			int maxNumWriteBuffers) {
+		super(exceptionHandler, "SortMerger spilling thread", dispatcher);
+		this.memManager = memManager;
+		this.ioManager = ioManager;
+		this.serializer = serializer;
+		this.comparator = comparator;
+		this.mergeReadMemory = sortReadMemory;
+		this.writeMemory = writeMemory;
+		this.maxFanIn = maxNumFileHandles;
+		this.spillChannelManager = spillingChannelManager;
+		this.largeRecordHandler = largeRecordHandler;
+		this.spillingBehaviour = spillingBehaviour;
+		this.minNumWriteBuffers = minNumWriteBuffers;
+		this.maxNumWriteBuffers = maxNumWriteBuffers;
+	}
+
+	/**
+	 * Entry point of the thread.
+	 */
+	public void go() throws IOException {
+
+		final Queue<CircularElement<E>> cache = new ArrayDeque<>();
+		boolean cacheOnly = readCache(cache);
+
+		// check whether the thread was canceled
+		if (!isRunning()) {
+			return;
+		}
+
+		MutableObjectIterator<E> largeRecords = null;
+
+		// check if we can stay in memory with the large record handler
+		if (cacheOnly && largeRecordHandler != null && largeRecordHandler.hasData()) {
+			List<MemorySegment> memoryForLargeRecordSorting = new ArrayList<>();
+
+			CircularElement<E> circElement;
+			while ((circElement = this.dispatcher.poll(SortStage.READ)) != null) {
+				circElement.getBuffer().dispose();
+				memoryForLargeRecordSorting.addAll(circElement.getMemory());
+			}
+
+			if (memoryForLargeRecordSorting.isEmpty()) {
+				cacheOnly = false;
+				LOG.debug("Going to disk-based merge because of large records.");
+			} else {
+				LOG.debug("Sorting large records, to add them to in-memory merge.");
+				largeRecords = largeRecordHandler.finishWriteAndSortKeys(memoryForLargeRecordSorting);
+			}
+		}
+
+		// ------------------- In-Memory Merge ------------------------
+		if (cacheOnly) {
+			mergeInMemory(cache, largeRecords);
+			return;
+		}
+
+		// ------------------- Spilling Phase ------------------------
+		List<ChannelWithBlockCount> channelIDs = startSpilling(cache);
+
+		// ------------------- Merging Phase ------------------------
+
+		mergeOnDisk(channelIDs);
+	}
+
+	@Override
+	public void close() throws InterruptedException {
+		super.close();
+		if (spillingBehaviourOpened) {
+			this.spillingBehaviour.close();
+			this.spillingBehaviourOpened = false;
+		}
+	}
+
+	private boolean readCache(Queue<CircularElement<E>> cache) throws IOException {
+		CircularElement<E> element;
+
+		// ------------------- In-Memory Cache ------------------------
+		// fill cache
+		while (isRunning()) {
+			// take next element from queue
+			try {
+				element = this.dispatcher.take(SortStage.SPILL);
```

Comment:
I though about it for a while after your comment. Even though we are slightly changing the behaviour I think it is actually fine. There are two situation where the thread can be interrupted.

1. When it is being closed via `ThreadBase#shutdown`. In this scenario both in the old and in the new code we are exiting the thread. In the new code we are throwing an exception, but it will be discarded in the `ThreadBase#internalExceptionHandler`.
2. If the thread is interrupted from outside. First of all I cannot think of exact scenario when this can happen. Nevertheless, I am pretty sure it should not happen during a successful processing and actually I doubt it is safe to continue on processing.